File: backend/domains/gateway/presentation/schemas/pricing.py
"""定价目录 API Schema。"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal
import uuid

from pydantic import BaseModel, ConfigDict, Field


class MoneyDisplaySchema(BaseModel):
    model_config = ConfigDict(strict=True)

    amount: str
    currency: Literal["CNY", "USD"]
    fx_rate_used: str


class UpstreamPricingUpsert(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")

    provider: str = Field(min_length=1, max_length=50)
    upstream_model: str = Field(min_length=1, max_length=200)
    capability: str = Field(default="chat", max_length=40)
    currency: Literal["CNY", "USD"] = "CNY"
    amount_per_million: dict[str, Decimal | None] = Field(
        description="input / output / cache_creation / cache_read 每 1M tokens 展示币金额"
    )


class LitellmUpstreamSyncRequest(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")

    providers: list[str] | None = Field(default=None, description="为空时使用已配置凭据提供商")


class EffectiveProviderResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    provider: str
    credential_count: int
    has_managed: bool
    has_user: bool


class UpstreamPricingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    provider: str
    upstream_model: str
    capability: str
    input_cost_per_token_usd: str
    output_cost_per_token_usd: str
    input_cost_per_million_display: MoneyDisplaySchema | None = None
    output_cost_per_million_display: MoneyDisplaySchema | None = None
    cache_creation_input_token_cost_usd: str | None = None
    cache_read_input_token_cost_usd: str | None = None
    effective_from: datetime
    effective_to: datetime | None
    version: int
    source: str
    display_currency: str | None = None
    fx_rate_used: str | None = None


class DownstreamPricingUpsert(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")

    scope: Literal["global", "team", "entitlement_plan"]
    scope_id: uuid.UUID | None = None
    gateway_model_id: uuid.UUID | None = None
    inheritance_strategy: Literal["mirror", "manual"] = "manual"
    currency: Literal["CNY", "USD"] = "CNY"
    amount_per_million: dict[str, Decimal | None] | None = None


class DownstreamPricingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    scope: str
    scope_id: uuid.UUID | None
    gateway_model_id: uuid.UUID | None
    inheritance_strategy: str
    input_cost_per_token_usd: str | None = None
    output_cost_per_token_usd: str | None = None
    input_cost_per_million_display: MoneyDisplaySchema | None = None
    output_cost_per_million_display: MoneyDisplaySchema | None = None
    effective_from: datetime
    effective_to: datetime | None
    version: int
    display_currency: str | None = None
    fx_rate_used: str | None = None


class PricingRateAdminView(BaseModel):
    """团队/平台管理员可见：含上游与毛利。"""

    model_config = ConfigDict(strict=True)

    gateway_model_id: uuid.UUID | None
    model_name: str | None
    downstream: DownstreamPricingResponse | None
    upstream: UpstreamPricingResponse | None
    margin_per_million_display: MoneyDisplaySchema | None = None
    hit_chain: list[str]


class PricingRateMemberView(BaseModel):
    """普通成员：仅下游生效价。"""

    model_config = ConfigDict(strict=True)

    gateway_model_id: uuid.UUID | None
    model_name: str | None
    input_cost_per_million_display: MoneyDisplaySchema | None = None
    output_cost_per_million_display: MoneyDisplaySchema | None = None
    inheritance_strategy: str | None = None
    display_currency: str


class SyncReportResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    created: int
    skipped: int


class PricingEstimateRequest(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")

    gateway_model_id: uuid.UUID
    input_tokens: int = Field(ge=0, default=0)
    output_tokens: int = Field(ge=0, default=0)
    cache_read_tokens: int = Field(ge=0, default=0)


class PricingEstimateResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    gateway_model_id: str
    hit_chain: list[str]
    upstream_cost_usd: str
    downstream_revenue_usd: str
    margin_usd: str
    rate_snapshot: dict[str, object]
    disclaimer: str


class UpstreamPricingAuditResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    models_without_upstream: list[str]
    upstream_without_model: list[str]
    registered_upstream_keys: int


class LitellmUpstreamSyncReportResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    created: int
    updated: int
    skipped_manual: int


class PricingReconciliationResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    team_id: str
    period: str
    requests: int
    cost_usd: str
    revenue_usd: str
    margin_usd: str
    top_models: list[dict[str, object]]


class FxRateResponse(BaseModel):
    model_config = ConfigDict(strict=True)

    usd_cny: str
    adapter: str = "static"
    default_display_currency: str


__all__ = [
    "DownstreamPricingResponse",
    "DownstreamPricingUpsert",
    "EffectiveProviderResponse",
    "FxRateResponse",
    "LitellmUpstreamSyncRequest",
    "MoneyDisplaySchema",
    "PricingRateAdminView",
    "PricingRateMemberView",
    "SyncReportResponse",
    "UpstreamPricingResponse",
    "UpstreamPricingUpsert",
]

File: frontend/src/api/gateway/pricing.ts
/**
 * AI Gateway · 计价目录（Pricing Catalog）
 *
 * 三层计价：
 * - 上游计价（Upstream）：成本侧，由 LiteLLM 价格表或手动维护
 * - 下游计价（Downstream）：收入侧，按 global / team / entitlement_plan 三种作用域
 * - 我的计价（My Prices）：当前调用方实际能看到的「单价 + 货币 + 继承策略」聚合视图
 *
 * 此外提供 FX 汇率、对账（reconciliation）、单次调用估价（estimate）、上游同步等运维端点。
 */

import { apiClient } from '@/api/client'
import type { DisplayCurrency, MoneyDisplay } from '@/types/money'

import { teamGatewayPath } from './_base'

/** GET /pricing/fx 汇率与适配器信息 */
export interface FxRateInfo {
  usd_cny: string
  adapter: string
  default_display_currency: string
}

// ---------- 上游计价 ----------

export interface UpstreamPricingRow {
  id: string
  provider: string
  upstream_model: string
  capability: string
  input_cost_per_token_usd: string
  output_cost_per_token_usd: string
  input_cost_per_million_display?: MoneyDisplay | null
  output_cost_per_million_display?: MoneyDisplay | null
  cache_creation_input_token_cost_usd?: string | null
  cache_read_input_token_cost_usd?: string | null
  effective_from: string
  effective_to: string | null
  version: number
  source: string
  display_currency?: string
  fx_rate_used?: string
}

export interface UpstreamPricingUpsertBody {
  provider: string
  upstream_model: string
  capability?: string
  currency?: DisplayCurrency
  /** 单位：每百万 token 单价（keys 同 backend `amount_per_million` 字段） */
  amount_per_million: Record<string, number | null>
}

// ---------- 下游计价 ----------

export interface DownstreamPricingRow {
  id: string
  scope: string
  scope_id: string | null
  gateway_model_id: string | null
  inheritance_strategy: string
  input_cost_per_token_usd?: string | null
  output_cost_per_token_usd?: string | null
  input_cost_per_million_display?: MoneyDisplay | null
  output_cost_per_million_display?: MoneyDisplay | null
  effective_from: string
  effective_to: string | null
  version: number
}

export interface DownstreamPricingUpsertBody {
  scope: 'global' | 'tenant' | 'entitlement_plan'
  scope_id?: string | null
  gateway_model_id?: string | null
  /** mirror：继承上游；manual：使用本行 amount_per_million */
  inheritance_strategy?: 'mirror' | 'manual'
  currency?: DisplayCurrency
  amount_per_million?: Record<string, number | null> | null
}

// ---------- 我的计价（聚合视图） ----------

export interface MyPriceRow {
  gateway_model_id: string | null
  model_name: string | null
  input_cost_per_million_display?: MoneyDisplay | null
  output_cost_per_million_display?: MoneyDisplay | null
  inheritance_strategy?: string | null
  display_currency: string
}

// ---------- 审计 / 同步 / 估价 / 对账 ----------

export interface UpstreamPricingAuditResult {
  models_without_upstream: string[]
  upstream_without_model: string[]
  registered_upstream_keys: number
}

export interface LitellmUpstreamSyncResult {
  created: number
  updated: number
  skipped_manual: number
}

export interface EffectiveProvider {
  provider: string
  credential_count: number
  has_managed: boolean
  has_user: boolean
}

export interface LitellmUpstreamSyncBody {
  providers?: string[] | null
}

export interface PricingEstimateBody {
  gateway_model_id: string
  input_tokens?: number
  output_tokens?: number
  cache_read_tokens?: number
}

export interface PricingEstimateResult {
  gateway_model_id: string
  /** 命中规则的链路（按优先级从高到低） */
  hit_chain: string[]
  upstream_cost_usd: string
  downstream_revenue_usd: string
  margin_usd: string
  rate_snapshot: Record<string, unknown>
  disclaimer: string
}

export interface PricingReconciliationResult {
  team_id: string
  period: string
  requests: number
  cost_usd: string
  revenue_usd: string
  margin_usd: string
  top_models: Array<{
    route_name: string | null
    requests: number
    cost_usd: string
    revenue_usd: string
    margin_usd: string
  }>
}

/** Pricing 资源 API */
export const pricingApi = {
  /** 获取当前 FX 汇率与适配器信息（首屏货币切换依赖） */
  getFxRates: (teamId: string) => apiClient.get<FxRateInfo>(teamGatewayPath(teamId, '/pricing/fx')),

  // --- 上游计价 ---
  /** 列出实际已配置凭据的 provider，用作上游同步白名单 */
  getEffectiveProviders: (teamId: string) =>
    apiClient.get<EffectiveProvider[]>(teamGatewayPath(teamId, '/pricing/effective-providers')),
  /** 列出上游计价（按 provider / currency 过滤） */
  listUpstreamPricing: (
    teamId: string,
    params?: { provider?: string; currency?: DisplayCurrency }
  ) => apiClient.get<UpstreamPricingRow[]>(teamGatewayPath(teamId, '/pricing/upstream'), params),
  /** 新增 / 更新一条上游计价（增量版本，旧版自动 effective_to 截断） */
  createUpstreamPricing: (teamId: string, body: UpstreamPricingUpsertBody) =>
    apiClient.post<UpstreamPricingRow>(teamGatewayPath(teamId, '/pricing/upstream'), body),
  /** 审计：列出缺失或多余的上游计价键 */
  auditUpstreamPricing: (teamId: string) =>
    apiClient.get<UpstreamPricingAuditResult>(teamGatewayPath(teamId, '/pricing/upstream/audit')),
  /** 从 LiteLLM 价格表同步上游计价（manual 行不会被覆盖） */
  syncUpstreamFromLitellm: (teamId: string, body?: LitellmUpstreamSyncBody) =>
    apiClient.post<LitellmUpstreamSyncResult>(
      teamGatewayPath(teamId, '/pricing/upstream/sync-from-litellm'),
      body ?? {}
    ),

  // --- 下游计价 ---
  /** 列出下游计价（按 scope / scope_id / currency 过滤） */
  listDownstreamPricing: (
    teamId: string,
    params?: {
      scope?: 'global' | 'tenant' | 'entitlement_plan'
      scope_id?: string
      currency?: DisplayCurrency
    }
  ) =>
    apiClient.get<DownstreamPricingRow[]>(teamGatewayPath(teamId, '/pricing/downstream'), params),
  /** 新增 / 更新一条下游计价 */
  createDownstreamPricing: (teamId: string, body: DownstreamPricingUpsertBody) =>
    apiClient.post<DownstreamPricingRow>(teamGatewayPath(teamId, '/pricing/downstream'), body),
  /**
   * 从上游同步下游计价（mirror 策略）。
   * - scope/scope_id 缺省时仅同步 global
   * - 返回新建条数与因冲突跳过的条数
   */
  syncDownstreamPricing: (
    teamId: string,
    params?: {
      scope?: 'tenant' | 'entitlement_plan'
      scope_id?: string
    }
  ) => {
    const qs = new URLSearchParams()
    if (params?.scope) qs.set('scope', params.scope)
    if (params?.scope_id) qs.set('scope_id', params.scope_id)
    const q = qs.toString()
    return apiClient.post<{ created: number; skipped: number }>(
      `${teamGatewayPath(teamId, '/pricing/downstream/sync')}${q ? `?${q}` : ''}`,
      {}
    )
  },

  // --- 我的计价 / 单次估价 / 对账 ---
  /** 当前调用方可见的「单价 + 货币 + 继承策略」聚合视图 */
  listMyPrices: (teamId: string, params?: { currency?: DisplayCurrency }) =>
    apiClient.get<MyPriceRow[]>(teamGatewayPath(teamId, '/pricing/my'), params),
  /** 单次调用估价（基于命中链路的 cost / revenue / margin） */
  estimatePricing: (teamId: string, body: PricingEstimateBody) =>
    apiClient.post<PricingEstimateResult>(teamGatewayPath(teamId, '/pricing/estimate'), body),
  /** 月度对账（团队维度，cost / revenue / margin / top models） */
  pricingReconciliation: (teamId: string, params: { year: number; month: number }) =>
    apiClient.get<PricingReconciliationResult>(
      teamGatewayPath(teamId, '/pricing/reconciliation'),
      params
    ),
} as const
